### PR TITLE
[GSoC 2019] Configure Web Notification Consumer

### LINKED ISSUE #374: [GSoC 2019] Store notifications in the database
- As discussed, the database schema needs to be updated for storing notifications of users. This issue will be an immediate follow up of #373. The notifications are supposed to be related to the Users table, so we should keep that in mind. 
- Also, DataFixtures are needed to be written for the above case.

### Tasks involved in this issue
- [x] Extend the database schema
- [x] Store Notifications in database
- [ ] Write DataFixtures for the above

### Further goals
 - This issue should deal with persisting the generated notification to the database so that they can be fetched to the UI afterward.

### PR DESCRIPTION
**Pull request by aquibbaig**
Fixes #374

### CODE REVIEW

File: site/src/Entity/Notification.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+{
+    /**
+     * The recipient of the Notification
+     *
+     * @var User
+     *
+     * @ORM\Column(name="recipient", nullable=false)
+     *
+     */
+    protected $recipient;
```
Comment (imphil) [inline site/src/Entity/Notification.php]: You don't need that, the recipient is in the notifiable annotation in mgilet-notificationbundle.

File: site/src/Consumer/WebNotificationConsumer.php [hunk 2]
```diff
@@ -15,27 +18,78 @@
...
-     * @param LoggerInterface $logger
+     * @param LoggerInterface     $logger
+     * @param NotificationManager $notificationManager
      */
-    public function __construct(LoggerInterface $logger)
-    {
+    public function __construct(
+        LoggerInterface $logger,
+        NotificationManager $notificationManager,
+        UserManagerInterface $userManager
```
Comment (imphil) [inline site/src/Consumer/WebNotificationConsumer.php]: this last argument is not documented in the phpdoc.

File: site/src/Consumer/WebNotificationConsumer.php [hunk 3]
```diff
@@ -15,27 +18,78 @@
...
+        $notification = new Notification();
+
+        // Populate the Notification Entity
+        $notification->setMessage($this->notification->getMessage());
+        $notification->setType($this->notification->getType());
+        $notification->setSubject($this->notification->getSubject());
+        $notification->setDate($this->notification->getCreatedAt());
+        $notification->setRecipient($this->notification->getRecipient());
+
+        if ($this->shouldHandle()) {
```
Comment (imphil) [inline site/src/Consumer/WebNotificationConsumer.php]: Why do you call that method here? It has already been called once we enter handle().

File: site/src/Consumer/WebNotificationConsumer.php [hunk 4]
```diff
@@ -15,27 +18,78 @@
...
 
-        return true;
+        $notification = new Notification();
+
+        // Populate the Notification Entity
+        $notification->setMessage($this->notification->getMessage());
+        $notification->setType($this->notification->getType());
+        $notification->setSubject($this->notification->getSubject());
+        $notification->setDate($this->notification->getCreatedAt());
+        $notification->setRecipient($this->notification->getRecipient());
```
Comment (imphil) [inline site/src/Consumer/WebNotificationConsumer.php]: The recipient is already given below in the notifiable part (the first parameter of addNotification()), so you don't need that here.

File: site/src/Consumer/WebNotificationConsumer.php [hunk 5]
```diff
@@ -15,27 +18,78 @@
...
+
+        if ($this->shouldHandle()) {
+            // Persist a Notification
+            $this->notificationManager->addNotification(array($this->notification->getRecipient()), $notification, true);
```
Comment (imphil) [inline site/src/Consumer/WebNotificationConsumer.php]: Please keep this line around 100 characters.

The newer way of writing array() is just using []

File: site/src/Consumer/WebNotificationConsumer.php (hunk 4 above)
Comment (aquibbaig) [reply site/src/Consumer/WebNotificationConsumer.php]: Agreed, But we are persisting the `$notification` variable which will not have a Recipient if we omit this line probably.

File: site/src/Consumer/WebNotificationConsumer.php (hunk 2 above)
Comment (aquibbaig) [reply site/src/Consumer/WebNotificationConsumer.php]: `userManager` is not required in this Consumer, yet

File: site/src/Entity/Notification.php (hunk 1 above)
Comment (aquibbaig) [reply site/src/Entity/Notification.php]: How can we fetch that while persisting the Notification for a specific user ?

Comment (aquibbaig) [reply site/src/Entity/Notification.php]: Cause not having a recipient while persisting crashes our consumer

File: site/src/Consumer/WebNotificationConsumer.php (hunk 4 above)
Comment (imphil) [reply site/src/Consumer/WebNotificationConsumer.php]: The recipient is the "notifiable entity", you don't need a recipient field in addition to that.

File: site/src/Entity/Notification.php (hunk 1 above)
Comment (imphil) [reply site/src/Entity/Notification.php]: You added an annotation to make the User entity a NotifableEntity; that's where the recipient of a message is stored in this case.

Comment (aquibbaig) [reply site/src/Entity/Notification.php]: That is correct, what was I even thinking!

**Comment by aquibbaig** [comment]
@imphil Does the `handle()` method need more info on the logger and a `try->catch` ?

**Pull request merged**